### PR TITLE
ORC: Implement TestGenericData and fix reader and writer issues

### PR DESCRIPTION
**Pull request by shardulm94**
This PR implements TestGenericData testing class for ORC and tries to fix issues with the ORC implementation uncovered through this test
Changes:
- Time is now represented with ORC long instead of int. According to Iceberg spec, time is stored with microsecond precision and int is not sufficient for this range (8.64e+10). 
- Timestamp with TZ is represented as "Timestamp Instant" ORC type, stored as UTC and retrieved as UTC as per spec.
- Several changes to GenericOrcReader and GenericOrcWriter to conform with the expected Java types
- I had to make the Iceberg ORC type attribute names (e.g `iceberg.binary-type`, `iceberg.length`) in ORCSchemaUtil public. I access them in GenericOrcReader & GenericOrcWriter to instantiate the correct readers/writers since the type information in TypeDescription without attributes is insufficient. A better way to handle this is would probably be to create OrcTypeDescriptionWithIcebergType visitor like we have for Avro and Parquet.

Unsure about:
- I am confused about what should be the right way to store Timestamp without TZ. ORC does have a timestamp type. Internally, it will store timestamp along with the local timezone and on the read side, ORC will perform offset calculation between reader and write timezone so that the resultant timestamp will be the same from the user's perspective. I have implemented the changes with this approach. However, another approach can be to just store the timestamp forced to the UTC timezone as a "Timestamp Instant" so that ORC does not apply any offset calculations on it and let Iceberg handle the read to LocalDateTime (I believe this is what we do for Avro Generics today).

cc: @edgarRd

### CODE REVIEW

File: data/src/main/java/org/apache/iceberg/data/orc/GenericOrcReader.java [hunk 1]
```diff
@@ -172,14 +208,13 @@ public Double convert(ColumnVector vector, int row) {
     }
   }
 
-  private static class TimestampConverter implements Converter<Long> {
-    private Long convert(TimestampColumnVector vector, int row) {
-      // compute microseconds past 1970.
-      return (vector.time[row] / 1000) * 1_000_000 + vector.nanos[row] / 1000;
+  private static class TimestampTzConverter implements Converter<OffsetDateTime> {
+    private OffsetDateTime convert(TimestampColumnVector vector, int row) {
+      return EPOCH.plus(vector.time[row], ChronoUnit.MILLIS).plus(vector.nanos[row] % 1_000_000, ChronoUnit.NANOS);
```
Comment (rdsr) [inline data/src/main/java/org/apache/iceberg/data/orc/GenericOrcReader.java]: Seems like `vector.nanos[row]` is not absolute time in nanos and is like `java.sql.Time.getNanos`. In that case we don't need to take the `%` ?

File: data/src/main/java/org/apache/iceberg/data/orc/GenericOrcReader.java [hunk 2]
```diff
@@ -284,9 +370,9 @@ public BigDecimal convert(ColumnVector vector, int row) {
       final int length = (int) vector.lengths[row];
 
       // serialize the keys
```
Comment (rdsr) [inline data/src/main/java/org/apache/iceberg/data/orc/GenericOrcReader.java]: Maybe take out this comment?

File: data/src/main/java/org/apache/iceberg/data/orc/GenericOrcWriter.java [hunk 3]
```diff
@@ -429,26 +537,51 @@ private static Converter buildConverter(TypeDescription schema) {
...
-        return new LongConverter();
+        ORCSchemaUtil.LongType longType = ORCSchemaUtil.LongType.valueOf(
+            schema.getAttributeValue(ORCSchemaUtil.ICEBERG_LONG_TYPE_ATTRIBUTE));
+        switch (longType) {
+          case TIME:
+            return new TimeConverter();
+          case LONG:
+            return new LongConverter();
+          default:
+            throw new IllegalStateException("Invalid Long type found in ORC type attribute");
```
Comment (rdsr) [inline data/src/main/java/org/apache/iceberg/data/orc/GenericOrcWriter.java]: nit: maybe log the long type?

File: data/src/main/java/org/apache/iceberg/data/orc/GenericOrcWriter.java [hunk 4]
```diff
@@ -429,26 +537,51 @@ private static Converter buildConverter(TypeDescription schema) {
...
+            schema.getAttributeValue(ORCSchemaUtil.ICEBERG_BINARY_TYPE_ATTRIBUTE));
+        switch (binaryType) {
+          case UUID:
+            return new UUIDConverter();
+          case FIXED:
+            return new FixedConverter();
+          case BINARY:
+            return new BytesConverter();
+          default:
+            throw new IllegalStateException("Invalid Binary type found in ORC type attribute");
```
Comment (rdsr) [inline data/src/main/java/org/apache/iceberg/data/orc/GenericOrcWriter.java]: log the invalid binary type?

**Comment by rdsr** [comment]
>However, another approach can be to just store the timestamp forced to the UTC timezone as a "Timestamp Instant" so that ORC does not apply any offset calculations on it and let Iceberg handle the read to LocalDateTime (I believe this is what we do for Avro Generics today).

I think though that we should handle both `Timestamp` and `TimestampTz` on the  read side, since in the future we could add a scenario where Iceberg could read ORC data not written by ORC writers.

When writing, I don't see a problem with always using the `TimestampTz` converter

File: data/src/main/java/org/apache/iceberg/data/orc/GenericOrcReader.java (hunk 1 above)
Comment (shardulm94) [reply data/src/main/java/org/apache/iceberg/data/orc/GenericOrcReader.java]: From https://issues.apache.org/jira/browse/ORC-546,
> Hive's TimestampColumnVector has a bad design with millis from 1970 and nanos within the second. This is consistent with java.sql.Timestamp, but it causes the millis to overlap with the nanos

Basically, the millis component of timestamp is double counted in both vector.time and also vector.nanos. Here we remove it from vector.nanos and only use it from vector.time

Comment (shardulm94) [reply data/src/main/java/org/apache/iceberg/data/orc/GenericOrcReader.java]: I also made a small change to not use millis from vector.time and rather use millis from vector.nanos. This is consistent with other places where we are handling ORC timestamps.

File: data/src/main/java/org/apache/iceberg/data/orc/GenericOrcReader.java (hunk 2 above)
Comment (shardulm94) [reply data/src/main/java/org/apache/iceberg/data/orc/GenericOrcReader.java]: Done

File: data/src/main/java/org/apache/iceberg/data/orc/GenericOrcWriter.java (hunk 3 above)
Comment (shardulm94) [reply data/src/main/java/org/apache/iceberg/data/orc/GenericOrcWriter.java]: Done

File: data/src/main/java/org/apache/iceberg/data/orc/GenericOrcWriter.java [hunk 5]
```diff
@@ -40,10 +50,13 @@
 import org.apache.orc.storage.ql.exec.vector.TimestampColumnVector;
 import org.apache.orc.storage.ql.exec.vector.VectorizedRowBatch;
 
+
 /**
```
[the remark this answers is not in the document]
Comment (shardulm94) [reply data/src/main/java/org/apache/iceberg/data/orc/GenericOrcWriter.java]: Removed empty message

File: orc/src/main/java/org/apache/iceberg/orc/ORCSchemaUtil.java [hunk 6]
```diff
@@ -70,20 +70,19 @@ public TypeDescription type() {
...
       ImmutableMap.<Type.TypeID, TypeDescription.Category>builder()
           .put(Type.TypeID.BOOLEAN, TypeDescription.Category.BOOLEAN)
           .put(Type.TypeID.INTEGER, TypeDescription.Category.INT)
-          .put(Type.TypeID.TIME, TypeDescription.Category.INT)
           .put(Type.TypeID.LONG, TypeDescription.Category.LONG)
+          .put(Type.TypeID.TIME, TypeDescription.Category.LONG)
           .put(Type.TypeID.FLOAT, TypeDescription.Category.FLOAT)
           .put(Type.TypeID.DOUBLE, TypeDescription.Category.DOUBLE)
           .put(Type.TypeID.DATE, TypeDescription.Category.DATE)
-          .put(Type.TypeID.TIMESTAMP, TypeDescription.Category.TIMESTAMP)
```
Comment (edgarRd) [inline orc/src/main/java/org/apache/iceberg/orc/ORCSchemaUtil.java]: What's `Type.TypeID.TIMESTAMP` now mapping to?

File: data/src/main/java/org/apache/iceberg/data/orc/GenericOrcWriter.java [hunk 7]
```diff
@@ -429,26 +534,51 @@ private static Converter buildConverter(TypeDescription schema) {
...
+        }
       case STRING:
       case CHAR:
       case VARCHAR:
         return new StringConverter();
       case DECIMAL:
-        return schema.getPrecision() <= 18 ?
-            new Decimal18Converter(schema) :
-            new Decimal38Converter(schema);
+        // TODO: Figure out the right cases in which to use the fastpath Decimal18Converter
```
Comment (edgarRd) [inline data/src/main/java/org/apache/iceberg/data/orc/GenericOrcWriter.java]: Should we create an issue to track this?

File: orc/src/main/java/org/apache/iceberg/orc/ORCSchemaUtil.java [hunk 8]
```diff
@@ -70,20 +70,19 @@ public TypeDescription type() {
   private static final String ICEBERG_ID_ATTRIBUTE = "iceberg.id";
   private static final String ICEBERG_REQUIRED_ATTRIBUTE = "iceberg.required";
 
-  private static final String ICEBERG_BINARY_TYPE_ATTRIBUTE = "iceberg.binary-type";
-  private static final String ICEBERG_INTEGER_TYPE_ATTRIBUTE = "iceberg.integer-type";
-  private static final String ICEBERG_FIELD_LENGTH = "iceberg.length";
+  public static final String ICEBERG_BINARY_TYPE_ATTRIBUTE = "iceberg.binary-type";
```
Comment (edgarRd) [inline orc/src/main/java/org/apache/iceberg/orc/ORCSchemaUtil.java]: Can we add a brief description javadoc on the now public constants?

File: orc/src/main/java/org/apache/iceberg/orc/ORCSchemaUtil.java (hunk 6 above)
Comment (shardulm94) [reply orc/src/main/java/org/apache/iceberg/orc/ORCSchemaUtil.java]: `Type.TypeID.TIMESTAMP` now can map to either `TypeDescription.Category.TIMESTAMP` or `TypeDescription.Category.TIMESTAMP_INSTANT` depending on with/without zone. The code for the same is present in `isSameType` where this map is being used.

File: orc/src/main/java/org/apache/iceberg/orc/ORCSchemaUtil.java (hunk 8 above)
Comment (shardulm94) [reply orc/src/main/java/org/apache/iceberg/orc/ORCSchemaUtil.java]: Done

File: data/src/main/java/org/apache/iceberg/data/orc/GenericOrcWriter.java (hunk 7 above)
Comment (shardulm94) [reply data/src/main/java/org/apache/iceberg/data/orc/GenericOrcWriter.java]: I just ended up fixing the issue in the latest commit. So should no longer be required.

**Comment by shardulm94** [comment]
This is ready for another round of review.

**Comment by rdsr** [comment]
Thanks @shardulm94 . I will review this today.

File: site/docs/spec.md [hunk 9]
```diff
@@ -483,25 +483,25 @@ Lists must use the [3-level representation](https://github.com/apache/parquet-fo
...
+| **`date`**         | `date`              |                                                                                         |
+| **`time`**         | `long`              | Stores microseconds from midnight.                                                      |
+| **`timestamp`**    | `timestamp`         |                                                                                         |
+| **`timestamptz`**  | `timestamp_instant` |                                       |
+| **`string`**       | `string`            | ORC `varchar` and `char` would also map to **`string`**.                                |
+| **`uuid`**         | `binary`            |                                                                                         |
+| **`fixed(L)`**     | `binary`            | The length would not be checked by the ORC reader and should be checked by the adapter. |
+| **`binary`**       | `binary`            |                                                                                         |
+| **`struct`**       | `struct`            | ORC `uniontype` would also map to **`struct`**.                                         |
```
Comment (rdsr) [inline site/docs/spec.md]: > ORC `uniontype` would also map to **`struct`**

Is this already supported in ORC?

File: data/src/test/java/org/apache/iceberg/data/orc/TestGenericData.java [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+      rows = Lists.newArrayList(reader);
+    }
+
+    for (int i = 0; i < expected.size(); i += 1) {
+      DataTestHelpers.assertEquals(schema.asStruct(), expected.get(i), rows.get(i));
+    }
+  }
+
+  @Test
+  public void writeAndValidateTimestamps() throws IOException {
```
Comment (rdsr) [inline data/src/test/java/org/apache/iceberg/data/orc/TestGenericData.java]: should this test go in the super class `DataTest`?

File: site/docs/spec.md (hunk 9 above)
Comment (shardulm94) [reply site/docs/spec.md]: I don't think so, atleast I don't see any code handling this. Not sure why it is mentioned in the spec.

Comment (rdsr) [reply site/docs/spec.md]: Makes sense to remove it, @edgarRd @rdblue  fyi

File: data/src/main/java/org/apache/iceberg/data/orc/GenericOrcReader.java [hunk 11]
```diff
@@ -189,30 +224,85 @@ public Long convert(ColumnVector vector, int row) {
...
     public String convert(ColumnVector vector, int row) {
       BinaryConverter converter = new BinaryConverter();
-      byte[] byteData = converter.convert(vector, row);
-      if (byteData == null) {
+      ByteBuffer byteBuffer = converter.convert(vector, row);
+      if (byteBuffer == null) {
         return null;
       }
-      return new String(byteData, StandardCharsets.UTF_8);
+      return new String(byteBuffer.array(), StandardCharsets.UTF_8);
```
Comment (rdblue) [inline data/src/main/java/org/apache/iceberg/data/orc/GenericOrcReader.java]: Any use of `ByteBuffer.array()` without `arrayOffset()`, `position()`, and `limit()` is concerning. Here, there is a strong dependence between this code and the behavior of the wrapped converter: it must almost always allocate a new `ByteBuffer` to return.

File: data/src/main/java/org/apache/iceberg/data/orc/GenericOrcWriter.java [hunk 12]
```diff
@@ -229,27 +293,68 @@ public void addValue(int rowId, byte[] data, ColumnVector output) {
...
+    public void addValue(int rowId, OffsetDateTime data, ColumnVector output) {
+      if (data == null) {
+        output.noNulls = false;
+        output.isNull[rowId] = true;
+      } else {
+        output.isNull[rowId] = false;
+        TimestampColumnVector cv = (TimestampColumnVector) output;
+        long micros = ChronoUnit.MICROS.between(EPOCH, data);
+        cv.time[rowId] = micros / 1_000; // millis
+        cv.nanos[rowId] = (int) (micros % 1_000_000) * 1_000; // nanos
```
Comment (rdblue) [inline data/src/main/java/org/apache/iceberg/data/orc/GenericOrcWriter.java]: Doesn't this report milliseconds twice?

```
OffsetDateTime now = OffsetDateTime.now(); // 2020-03-17T00:21:46.844Z
long micros = ChronoUnit.MICROS.between(EPOCH, now); // 1584404506844000
long millis = micros / 1_000; // 1584404506844 <--- note 844 is millis
long nanos = (micros % 1_000_000) * 1_000; // 844000000 <--- millis are included
```

Does the millis/nanos format discard digits above 1,000,000 by convention? (ping @omalley)

Comment (shardulm94) [reply data/src/main/java/org/apache/iceberg/data/orc/GenericOrcWriter.java]: Yes we report the millis twice. ORC code only keeps seconds from cv.time and so handles this transparently. Since ORC expects the double reporting, I think it should be fine if we keep double reporting. @omalley Thoughts?

Note that the ORC reader double reports millis even if the writer does not double report millis.  Details in https://issues.apache.org/jira/browse/ORC-546

**Comment by shardulm94** [comment]
Sorry this took so long to address. Got sidetracked. I made a few changes in the latest commit.
- Addressed the PR comments
- We now use ZoneId and ZonedDateTime instead of ZoneOffset and OffsetDateTime. So this works now with timezones which require DST adjustments.
- Rebased to latest master

File: data/src/main/java/org/apache/iceberg/data/orc/GenericOrcReader.java (hunk 11 above)
Comment (shardulm94) [reply data/src/main/java/org/apache/iceberg/data/orc/GenericOrcReader.java]: I replaced this by not using ByteBuffers and directly creating a String from the vectors returned by ORC.

File: data/src/test/java/org/apache/iceberg/data/orc/TestGenericData.java (hunk 10 above)
Comment (shardulm94) [reply data/src/test/java/org/apache/iceberg/data/orc/TestGenericData.java]: Today the Parquet implementation does not support Timestamp without Timezone. So this would fail if added to the super class. Can we address this in a separate issue?

File: site/docs/spec.md (hunk 9 above)
Comment (shardulm94) [reply site/docs/spec.md]: Done

**Comment by rdblue** [comment]
> We now use ZoneId and ZonedDateTime instead of ZoneOffset and OffsetDateTime

Avro and Parquet currently return OffsetDateTime and LocalDateTime for timestamptz and timestamp. The values returned for all 3 formats should match so that users can expect consistent classes. So we should either use OffsetDateTime, or change all 3 to use ZonedDateTime.

I think we should continue to use OffsetDateTime, but I'm up for hearing why we should use ZonedDateTime. The rationale for OffsetDateTime is that we don't store zone information. All timestamps are converted to an instant that is the same across all zones. We pass that instant back as a UTC value for convenience, but Iceberg doesn't interpret that instant with respect to some zone. Conversions are the responsibility of processing engines, Iceberg just stores values and returns them unmodified.

**Comment by shardulm94** [comment]
@rdblue The use case is a bit different here. The end user Java object is still LocalDateTime for timestamp and OffsetDateTime for timestamptz. The timestamptz implementation for ORC still uses OffsetDateTime internally and is consistent with Avro and Parquet. I perhaps should have made the comment more clearer.

This is in relation to timestamp. In Avro and Parquet we model LocalDateTime as an instant and store it as int64 then reverse it on the read side to get OffsetDateTime and strip out the UTC zone to get back LocalDateTime. ORC already has support for Timestamp; check "timestamp with local time zone" in https://orc.apache.org/docs/types.html at the end. You basically provide ORC time since UTC epoch in the current writer timezone, ORC will store the time and writer timezone. During reads, ORC will compare the reader and writer timezone and apply the necessary offset to return time since UTC epoch in the reader timezone which would be equivalent to the local time on the writers side.

So here, we use ZonedDateTime to convert LocalDateTime into the "writer timezone specific time" before passing to ORC. OffsetDateTime wont work here because the "writer timezone specific time" is sensitive to the DST rules within the timezone and so the offset may not be constant for all dates.

**Comment by rdblue** [comment]
Avro and Parquet use OffsetDateTime just to get the microseconds from epoch. It isn't changing the value at all. It looks like ZoneDateTime code here relies on the environment and modifies the value -- I don't think that is correct. If there is any situation where a combination of environment time zones modifies the value between write and read, it is incorrect.

**Comment by shardulm94** [comment]
The code internally does modify the value between timezones, but it does it in such a way that the end result (LocalDateTime) is the same. However if that is not acceptable, then we cannot use the Timestamp type from ORC. We can use TimestampInstant type from ORC and we can then store micros from epoch assuming UTC just like we do for Avro and Parquet. ORC will not perform any offset manipulation for TimestampInstant.

**Comment by rdblue** [comment]
> It does depend the environment time zones though. If that is not acceptable, then we cannot use the Timestamp type from ORC.

Yeah, that doesn't work. Storage (Iceberg and file formats) should not be responsible for modifying values and it should never be possible to get a different answer if the environment time zone changes between read and write time. Iceberg always returns the exact same value on read that was passed in on write, but it also keeps enough metadata that the engine can decide whether to transform the value to display to the user (adjust-to-utc settings in Parquet and Avro).

From the [ORC docs](https://orc.apache.org/docs/types.html#timestamps), timestamp with local time zone "does change based on the time zone of the reader" so we should never use it. Instead, let's use regular timestamp with a UTC value for both Iceberg `timestamp` and `timestamptz`. We should also add metadata to the timestamp types to encode `adjust-to-utc` like Parquet and Avro. `timestamp` uses `adjust-to-utc=false` and `timestamptz` uses `adjust-to-utc=true`.

File: data/src/main/java/org/apache/iceberg/data/orc/GenericOrcReader.java [hunk 13]
```diff
@@ -189,30 +226,89 @@ public Long convert(ColumnVector vector, int row) {
...
+    @Override
+    public ByteBuffer convert(ColumnVector vector, int row) {
+      int rowIndex = vector.isRepeating ? 0 : row;
+      if (!vector.noNulls && vector.isNull[rowIndex]) {
+        return null;
+      } else {
+        BytesColumnVector bytesVector = (BytesColumnVector) vector;
+        ByteBuffer buf = ByteBuffer.allocate(bytesVector.length[rowIndex]);
+        buf.put(bytesVector.vector[rowIndex], bytesVector.start[rowIndex], bytesVector.length[rowIndex]);
+        buf.rewind();
```
Comment (rdblue) [inline data/src/main/java/org/apache/iceberg/data/orc/GenericOrcReader.java]: This should not copy. Instead, it should point a `ByteBuffer` at the underlying bytes in the vector using `ByteBuffer.wrap(array, offset, length)`.

File: data/src/main/java/org/apache/iceberg/data/orc/GenericOrcReader.java [hunk 14]
```diff
@@ -236,11 +332,7 @@ public BigDecimal convert(ColumnVector vector, int row) {
       Preconditions.checkArgument(icebergField.type().isListType());
       TypeDescription child = schema.getChildren().get(0);
 
-      childConverter = buildConverter(icebergField
-          .type()
-          .asListType()
-          .fields()
-          .get(0), child);
+      childConverter = buildConverter(icebergField.type().asListType().fields().get(0), child);
```
Comment (rdblue) [inline data/src/main/java/org/apache/iceberg/data/orc/GenericOrcReader.java]: I prefer this style, but it's better not to make non-functional changes like this to avoid commit conflicts.

File: data/src/main/java/org/apache/iceberg/data/orc/GenericOrcReader.java [hunk 15]
```diff
@@ -351,20 +441,43 @@ private static Converter buildConverter(final Types.NestedField icebergField,
       case SHORT:
         return new ShortConverter();
       case DATE:
+        return new DateConverter();
       case INT:
         return new IntConverter();
       case LONG:
-        return new LongConverter();
+        ORCSchemaUtil.LongType longType =
+            ORCSchemaUtil.LongType.valueOf(schema.getAttributeValue(ORCSchemaUtil.ICEBERG_LONG_TYPE_ATTRIBUTE));
```
Comment (rdblue) [inline data/src/main/java/org/apache/iceberg/data/orc/GenericOrcReader.java]: Is the spec up to date with these ORC metadata fields?

File: data/src/main/java/org/apache/iceberg/data/orc/GenericOrcWriter.java (hunk 12 above)
Comment (rdblue) [reply data/src/main/java/org/apache/iceberg/data/orc/GenericOrcWriter.java]: Let's make sure this behavior is clear in the spec.

File: data/src/test/java/org/apache/iceberg/data/DataTest.java [hunk 16]
```diff
@@ -48,7 +48,7 @@
       optional(105, "f", Types.FloatType.get()),
       required(106, "d", Types.DoubleType.get()),
       optional(107, "date", Types.DateType.get()),
-      required(108, "ts", Types.TimestampType.withZone()),
+      required(108, "tsTz", Types.TimestampType.withZone()),
```
Comment (rdblue) [inline data/src/test/java/org/apache/iceberg/data/DataTest.java]: Is this necessary?

File: data/src/test/java/org/apache/iceberg/data/orc/TestGenericData.java (hunk 10 above)
Comment (rdblue) [reply data/src/test/java/org/apache/iceberg/data/orc/TestGenericData.java]: Yeah, we need to update Parquet now that the new logical types API is out. (FYI @Fokko)

File: orc/src/main/java/org/apache/iceberg/orc/ORCSchemaUtil.java [hunk 17]
```diff
@@ -41,12 +41,12 @@
  */
 public final class ORCSchemaUtil {
 
-  private enum BinaryType {
+  public enum BinaryType {
```
Comment (rdblue) [inline orc/src/main/java/org/apache/iceberg/orc/ORCSchemaUtil.java]: Why is this now public?

File: site/docs/spec.md [hunk 18]
```diff
@@ -651,5 +651,3 @@ This serialization scheme is for storing single values as individual binary valu
 | **`struct`**                 | Not supported                                                                                                |
 | **`list`**                   | Not supported                                                                                                |
 | **`map`**                    | Not supported                                                                                                |
-
-
```
Comment (rdblue) [inline site/docs/spec.md]: Can you remove unnecessary whitespace changes to reduce commit problems?

**Comment by rdblue** [comment]
Thanks, @shardulm94! I did a more thorough check. Looks pretty good other than the timestamp issue.

**Comment by omalley** [comment]
I think the current usage of the java types is misleading. I think it should be:
- timestamp = LocalDateTime
- timestampz = Instant

I also find the difference between fixed & unfixed binary to be unexpected. I would use ByteBuffer for both of them, instead of byte[] for fixed.

**Comment by rdblue** [comment]
For Avro and Parquet, `timestamp` is returned as `LocalDateTime`, so I agree that is what ORC should do.

Avro and Parquet use `OffsetDateTime` for `timestamptz` with offset +00:00. That makes it easy to work with the value, but ensures it is always equivalent to an `Instant`. I'd be fine with either one, but since we already produce `OffsetDateTime` I'd prefer to start with what the other formats already do rather than changing all 3 and the expected type produced by the generics API.

**Comment by rdblue** [comment]
Forgot to add: I agree it would make more sense to use `ByteBuffer` for fixed, at least in the read path. Maybe we can add an option as a follow-up to configure that.

**Comment by omalley** [comment]
Ok, I have some suggestions for the timestamp stuff. Most of it boils down to using the useUtc flag, which makes life much easier. I've put my changes into https://github.com/omalley/iceberg/commit/d6b785413ac4f3c93a280229c28f147821b45168. Let me know if I should open a separate PR.

File: data/src/main/java/org/apache/iceberg/data/orc/GenericOrcReader.java (hunk 14 above)
Comment (shardulm94) [reply data/src/main/java/org/apache/iceberg/data/orc/GenericOrcReader.java]: Reverted

File: data/src/main/java/org/apache/iceberg/data/orc/GenericOrcReader.java (hunk 13 above)
Comment (shardulm94) [reply data/src/main/java/org/apache/iceberg/data/orc/GenericOrcReader.java]: Done

File: data/src/main/java/org/apache/iceberg/data/orc/GenericOrcReader.java (hunk 15 above)
Comment (shardulm94) [reply data/src/main/java/org/apache/iceberg/data/orc/GenericOrcReader.java]: Updated the spec now

File: data/src/main/java/org/apache/iceberg/data/orc/GenericOrcWriter.java (hunk 12 above)
Comment (shardulm94) [reply data/src/main/java/org/apache/iceberg/data/orc/GenericOrcWriter.java]: Added this behaviour as a note for the timestamp types in the ORC type mapping in the spec

File: data/src/test/java/org/apache/iceberg/data/DataTest.java (hunk 16 above)
Comment (shardulm94) [reply data/src/test/java/org/apache/iceberg/data/DataTest.java]: Reverted

File: orc/src/main/java/org/apache/iceberg/orc/ORCSchemaUtil.java (hunk 17 above)
Comment (shardulm94) [reply orc/src/main/java/org/apache/iceberg/orc/ORCSchemaUtil.java]: In the readers and writers, e.g. GenericOrcReader and GenericOrcWriter, I would like to instantiate different converters for fixed v/s binary v/s uuid. Making this public will allow us to reference this enum while parsing the ORC type attribute in those classes. I think it might be cleaner to use a visitor similar to AvroSchemaWithTypeVisitor in GenericOrcReader/Writer so that we don't need to reference these classes there as we get enough information from iceberg types to disambiguate. But the ORC implementation today lacks such visitors.

File: site/docs/spec.md (hunk 18 above)
Comment (shardulm94) [reply site/docs/spec.md]: Done

File: site/docs/spec.md [hunk 19]
```diff
@@ -486,26 +486,29 @@ Lists must use the [3-level representation](https://github.com/apache/parquet-fo
...
+| **`long`**         | `long`              |                                                      |                                                                                         |
+| **`float`**        | `float`             |                                                      |                                                                                         |
+| **`double`**       | `double`            |                                                      |                                                                                         |
+| **`decimal(P,S)`** | `decimal`           |                                                      |                                                                                         |
+| **`date`**         | `date`              |                                                      |                                                                                         |
+| **`time`**         | `long`              | `iceberg.long-type`=`DATE`                           | Stores microseconds from midnight.                                                      |
```
Comment (rdblue) [inline site/docs/spec.md]: Looks like this should be `time` and not `date`?

Comment (shardulm94) [reply site/docs/spec.md]: My bad. Fixed.

**Comment by rdblue** [comment]
Thanks @shardulm94, I merged this!

File: data/src/test/java/org/apache/iceberg/data/orc/TestGenericData.java (hunk 10 above)
Comment (rdsr) [reply data/src/test/java/org/apache/iceberg/data/orc/TestGenericData.java:77]: Make sense!
